include/crow/websocket.h

```diff
@@ -120,28 +120,28 @@ namespace crow
 
             /// Send data through the socket.
             template<typename CompletionHandler>
-            void dispatch(CompletionHandler handler)
+            void dispatch(CompletionHandler&& handler)
             {
-                adaptor_.get_io_service().dispatch(handler);
+                adaptor_.get_io_service().dispatch(std::forward<CompletionHandler>(handler));
             }
 
             /// Send data through the socket and return immediately.
             template<typename CompletionHandler>
-            void post(CompletionHandler handler)
+            void post(CompletionHandler&& handler)
             {
-                adaptor_.get_io_service().post(handler);
+                adaptor_.get_io_service().post(std::forward<CompletionHandler>(handler));
             }
 
             /// Send a "Ping" message.
 
             ///
             /// Usually invoked to check if the other point is still online.
-            void send_ping(const std::string& msg) override
+            void send_ping(std::string msg) override
             {
-                dispatch([this, msg] {
+                dispatch([this, msg = std::move(msg)]() mutable {
                     auto header = build_header(0x9, msg.size());
```

Comment:
Crow still supports C++11. Unfortunately, move capture was added only in c++14.

We use `#ifdef CROW_CAN_USE_CPP14` to handle such cases. Take a look at the routing.h as an example

Reply:
Oops. I didn't know that was c++14 feature ... well, I'll make small update for this concept to work.

And some other question ... How `io_service().dispatch()` handles received event instance, by moving, or copying? If it copies received handler, than we may remove that copy by using shared_ptr for every `send_data` request.

This makes copy operation of newly created lambda object relatively trivial, however, it introduces unnecessary allocation if given string is smaller than SBO size. What do you think about this idea?

```
                auto p_msg = std::make_shared<std::string>(std::move(msg));
                dispatch([this, p_msg]() mutable {
                    auto header = build_header(0x9, p_msg->size());
                    write_buffers_.emplace_back(std::move(header));
                    write_buffers_.emplace_back(std::move(*p_msg));
                    do_write();
                });
```

Reply:
For new commit 99217d8b4076bf9d67de14c3861d29c259887d0d, I have extracted lambda capture as explicit structure, and took advantage of move sementics by moving given struct. It still takes all advantages of originally proposed, however, It seems `SendMessageType` as move-only type makes program un-compillable. It seems still there's copy operation exists. 

If we use `asio::post()` instead of `get_io_executor().post()`, move-only object become compilable(in commit b6ec1bc540c643fa7ff6bf71f017d2a08854d0e6). In this way, we don't have to create `shared_ptr` everytime to avoid copy.